File: 프로그래머스/2/42883. 큰 수 만들기/큰 수 만들기.java
import java.util.*;

public class Solution {
    public String solution(String number, int k) {
        Deque<Character> stack = new ArrayDeque<>();

        for (char digit : number.toCharArray()) {
            while (!stack.isEmpty() && k > 0 && stack.peekLast() < digit) {
                stack.pollLast();
                k--;
            }
            stack.offerLast(digit);
        }

        while (k-- > 0) {
            stack.pollLast();
        }

        StringBuilder sb = new StringBuilder();
        for (char ch : stack) {
            sb.append(ch);
        }

        return sb.toString();
    }
}

File: 프로그래머스/2/42883. 큰 수 만들기/큰 수 만들기.cpp
#include <bits/stdc++.h>
using namespace std;

string solution(string number, int k) {
    string res = number;
    auto it = res.begin();
    
    while (k > 0) {
        if (it + 1 != res.end() && *it >= *(it + 1)) {
            it++;
        }
        else {
            res.erase(it);
            k--;
            it = res.begin();
        }
    }
    return res;
}